File: ipad/edf_reader.py
"""Pure-Python EDF file reader — no C extensions required.

Reads EDF/EDF+ files using only struct and numpy, making it compatible
with iPad Python environments (Pyto) where C-extension packages like
pyedflib and MNE cannot be installed.

Based on the European Data Format specification:
https://www.edfplus.info/specs/edf.html
"""

import struct
import numpy as np

# Old 10-20 nomenclature to new (ACNS 2006 standard)
OLD_TO_NEW = {"T3": "T7", "T4": "T8", "T5": "P7", "T6": "P8"}


def read_edf(file_path):
    """Read an EDF file and return channel data, names, and sample rate.

    Args:
        file_path: Path to the .edf file.

    Returns:
        dict with keys:
            'data': np.ndarray of shape (n_channels, n_samples) in Volts
            'channel_names': list of channel name strings
            'sfreq': sampling frequency in Hz
            'duration': total duration in seconds
            'n_channels': number of channels
            'patient_info': dict with patient metadata from header
    """
    with open(file_path, "rb") as f:
        # --- HEADER (256 bytes) ---
        version = f.read(8).decode("ascii").strip()
        patient_id = f.read(80).decode("ascii").strip()
        recording_id = f.read(80).decode("ascii").strip()
        start_date = f.read(8).decode("ascii").strip()
        start_time = f.read(8).decode("ascii").strip()
        header_bytes = int(f.read(8).decode("ascii").strip())
        reserved = f.read(44).decode("ascii").strip()
        n_data_records = int(f.read(8).decode("ascii").strip())
        record_duration = float(f.read(8).decode("ascii").strip())
        n_channels = int(f.read(4).decode("ascii").strip())

        # --- CHANNEL HEADERS (256 bytes per channel) ---
        labels = [f.read(16).decode("ascii").strip() for _ in range(n_channels)]
        transducers = [f.read(80).decode("ascii").strip() for _ in range(n_channels)]
        phys_dims = [f.read(8).decode("ascii").strip() for _ in range(n_channels)]
        phys_mins = [float(f.read(8).decode("ascii").strip()) for _ in range(n_channels)]
        phys_maxs = [float(f.read(8).decode("ascii").strip()) for _ in range(n_channels)]
        dig_mins = [float(f.read(8).decode("ascii").strip()) for _ in range(n_channels)]
        dig_maxs = [float(f.read(8).decode("ascii").strip()) for _ in range(n_channels)]
        prefilters = [f.read(80).decode("ascii").strip() for _ in range(n_channels)]
        samples_per_record = [int(f.read(8).decode("ascii").strip()) for _ in range(n_channels)]
        ch_reserved = [f.read(32).decode("ascii").strip() for _ in range(n_channels)]

        # Compute scaling factors: digital -> physical (Volts)
        # physical = (digital - dig_min) / (dig_max - dig_min) * (phys_max - phys_min) + phys_min
        scales = []
        offsets = []
        for i in range(n_channels):
            dig_range = dig_maxs[i] - dig_mins[i]
            phys_range = phys_maxs[i] - phys_mins[i]
            if dig_range == 0:
                scales.append(0.0)
                offsets.append(0.0)
            else:
                scale = phys_range / dig_range
                offset = phys_mins[i] - dig_mins[i] * scale
                scales.append(scale)
                offsets.append(offset)

        # --- DATA RECORDS ---
        # EDF stores data as interleaved records:
        # [record1_ch1, record1_ch2, ..., record2_ch1, record2_ch2, ...]
        all_data = []
        for _ in range(n_data_records):
            record_channels = []
            for ch_idx in range(n_channels):
                n_samps = samples_per_record[ch_idx]
                raw_bytes = f.read(n_samps * 2)  # 16-bit integers
                digital = np.frombuffer(raw_bytes, dtype="<i2")
                # Convert to physical units
                physical = digital.astype(np.float64) * scales[ch_idx] + offsets[ch_idx]
                record_channels.append(physical)
            all_data.append(record_channels)

    # Concatenate records for each channel
    data = np.zeros((n_channels, n_data_records * samples_per_record[0]))
    for ch_idx in range(n_channels):
        segments = [all_data[rec][ch_idx] for rec in range(n_data_records)]
        data[ch_idx] = np.concatenate(segments)

    # Clean up channel names: strip "EEG " prefix, rename old nomenclature
    clean_names = []
    for label in labels:
        name = label
        # Strip EEG prefix
        for prefix in ["EEG ", "EEG-", "EEG"]:
            if name.startswith(prefix) and len(name) > len(prefix):
                name = name[len(prefix):].strip()
                break
        # Rename old 10-20 names
        name = OLD_TO_NEW.get(name, name)
        clean_names.append(name)

    # Convert physical units to Volts if needed
    # EDF physical dimension is usually "uV" for EEG
    for i in range(n_channels):
        unit = phys_dims[i].lower().strip()
        if unit in ("uv", "µv", "microvolt", "microvolts"):
            data[i] *= 1e-6  # Convert µV to V

    sfreq = samples_per_record[0] / record_duration
    duration = n_data_records * record_duration

    return {
        "data": data,
        "channel_names": clean_names,
        "sfreq": sfreq,
        "duration": duration,
        "n_channels": n_channels,
        "patient_info": {
            "patient_id": patient_id,
            "recording_id": recording_id,
            "start_date": start_date,
            "start_time": start_time,
        },
    }

File: EEGViewer/EEGViewer/Models/EDFReader.swift
// EDFReader.swift
// Pure-Swift EDF file parser — reads EDF/EDF+ files using Foundation Data.
// Port of ipad/edf_reader.py, following the EDF specification:
// https://www.edfplus.info/specs/edf.html

import Foundation
import Accelerate

enum EDFError: Error, LocalizedError {
    case fileNotFound
    case invalidHeader(String)
    case dataTruncated

    var errorDescription: String? {
        switch self {
        case .fileNotFound: return "EDF file not found"
        case .invalidHeader(let msg): return "Invalid EDF header: \(msg)"
        case .dataTruncated: return "EDF data is truncated"
        }
    }
}

struct EDFReader {

    /// Read an EDF file from disk and return parsed data.
    static func read(url: URL) throws -> EDFData {
        guard let fileData = try? Data(contentsOf: url) else {
            throw EDFError.fileNotFound
        }
        return try parse(data: fileData)
    }

    /// Parse EDF data from a Data buffer.
    static func parse(data: Data) throws -> EDFData {
        var offset = 0

        // --- MAIN HEADER (256 bytes) ---
        let _ = readASCII(data, offset: &offset, length: 8)           // version
        let patientID = readASCII(data, offset: &offset, length: 80)
        let recordingID = readASCII(data, offset: &offset, length: 80)
        let startDate = readASCII(data, offset: &offset, length: 8)
        let startTime = readASCII(data, offset: &offset, length: 8)
        let _ = readASCII(data, offset: &offset, length: 8)           // header bytes
        let _ = readASCII(data, offset: &offset, length: 44)          // reserved

        guard let nDataRecords = Int(readASCII(data, offset: &offset, length: 8)) else {
            throw EDFError.invalidHeader("Cannot parse number of data records")
        }
        guard let recordDuration = Float(readASCII(data, offset: &offset, length: 8)) else {
            throw EDFError.invalidHeader("Cannot parse record duration")
        }
        guard let nChannels = Int(readASCII(data, offset: &offset, length: 4)) else {
            throw EDFError.invalidHeader("Cannot parse number of channels")
        }

        // --- CHANNEL HEADERS (256 bytes per channel) ---
        var labels = [String]()
        for _ in 0..<nChannels {
            labels.append(readASCII(data, offset: &offset, length: 16))
        }

        // Transducers (80 bytes each) — skip
        offset += nChannels * 80

        var physDims = [String]()
        for _ in 0..<nChannels {
            physDims.append(readASCII(data, offset: &offset, length: 8))
        }

        var physMins = [Float]()
        for _ in 0..<nChannels {
            physMins.append(Float(readASCII(data, offset: &offset, length: 8)) ?? 0)
        }
        var physMaxs = [Float]()
        for _ in 0..<nChannels {
            physMaxs.append(Float(readASCII(data, offset: &offset, length: 8)) ?? 0)
        }
        var digMins = [Float]()
        for _ in 0..<nChannels {
            digMins.append(Float(readASCII(data, offset: &offset, length: 8)) ?? 0)
        }
        var digMaxs = [Float]()
        for _ in 0..<nChannels {
            digMaxs.append(Float(readASCII(data, offset: &offset, length: 8)) ?? 0)
        }

        // Prefilters (80 bytes each) — skip
        offset += nChannels * 80

        var samplesPerRecord = [Int]()
        for _ in 0..<nChannels {
            samplesPerRecord.append(Int(readASCII(data, offset: &offset, length: 8)) ?? 0)
        }

        // Channel reserved (32 bytes each) — skip
        offset += nChannels * 32

        // --- Compute scaling factors ---
        var scales = [Float](repeating: 0, count: nChannels)
        var offsets = [Float](repeating: 0, count: nChannels)
        for i in 0..<nChannels {
            let digRange = digMaxs[i] - digMins[i]
            let physRange = physMaxs[i] - physMins[i]
            if digRange == 0 {
                scales[i] = 0
                offsets[i] = 0
            } else {
                scales[i] = physRange / digRange
                offsets[i] = physMins[i] - digMins[i] * scales[i]
            }
        }

        // --- DATA RECORDS ---
        let totalSamplesPerChannel = nDataRecords * samplesPerRecord[0]
        var channelData = [[Float]](repeating: [Float](repeating: 0, count: totalSamplesPerChannel), count: nChannels)

        data.withUnsafeBytes { rawBuffer in
            var dataOffset = offset
            for rec in 0..<nDataRecords {
                for ch in 0..<nChannels {
                    let nSamps = samplesPerRecord[ch]
                    let baseIdx = rec * nSamps
                    let bytesNeeded = nSamps * 2

                    guard dataOffset + bytesNeeded <= data.count else { return }

                    // Read Int16 values and convert to Float with scaling
                    let int16Ptr = rawBuffer.baseAddress!.advanced(by: dataOffset)
                        .assumingMemoryBound(to: Int16.self)

                    for s in 0..<nSamps {
                        let digital = Float(int16Ptr[s])
                        channelData[ch][baseIdx + s] = digital * scales[ch] + offsets[ch]
                    }

                    dataOffset += bytesNeeded
                }
            }
        }

        // --- Clean channel names ---
        var cleanNames = [String]()
        for label in labels {
            var name = label
            // Strip EEG prefix
            for prefix in ["EEG ", "EEG-", "EEG"] {
                if name.hasPrefix(prefix) && name.count > prefix.count {
                    name = String(name.dropFirst(prefix.count)).trimmingCharacters(in: .whitespaces)
                    break
                }
            }
            // Rename old 10-20 names
            if let newName = Constants.oldToNewNames[name] {
                name = newName
            }
            cleanNames.append(name)
        }

        // --- Convert units to Volts ---
        for i in 0..<nChannels {
            let unit = physDims[i].lowercased().trimmingCharacters(in: .whitespaces)
            if ["uv", "µv", "microvolt", "microvolts"].contains(unit) {
                // Convert µV to V
                var scale: Float = 1e-6
                vDSP_vsmul(channelData[i], 1, &scale, &channelData[i], 1, vDSP_Length(channelData[i].count))
            }
        }

        let sfreq = Float(samplesPerRecord[0]) / recordDuration
        let duration = Float(nDataRecords) * recordDuration

        return EDFData(
            channelNames: cleanNames,
            data: channelData,
            sfreq: sfreq,
            duration: duration,
            nChannels: nChannels,
            patientInfo: EDFData.PatientInfo(
                patientID: patientID,
                recordingID: recordingID,
                startDate: startDate,
                startTime: startTime
            )
        )
    }

    // MARK: - Helpers

    /// Read ASCII string from Data at offset, trimming whitespace. Advances offset.
    private static func readASCII(_ data: Data, offset: inout Int, length: Int) -> String {
        let end = min(offset + length, data.count)
        let slice = data[offset..<end]
        offset = end
        return String(data: slice, encoding: .ascii)?
            .trimmingCharacters(in: .whitespaces) ?? ""
    }
}
